Add a chapter on divergence

jackh726 opened the pull request:
It was little tricky when trying to describe *diverging blocks*. The compiler's implementation maintains sort of a "global state" when checking an expression and sub-expressions, which it resets on conditional things. Semantically, I think the way I worded it is much clearer than trying to match the implementation.

Happy to hear any specific feedback, Lcnr made did an initial review pass in https://github.com/rust-lang/project-goal-reference-expansion/commit/407917146a090fac7ec3ae31132447bce8c98137, so the second commit tries to address that.

Closes https://github.com/rust-lang/reference/issues/271
Closes https://github.com/rust-lang/reference/issues/1033

jackh726 committed:
Write chapter on divergence

jackh726 committed:
Slight edits from lcnr's review

jackh726 commented:
Thanks for the review @traviscross. Good points here, I'll work on sorting through them today/tomorrow.

Happy to jump on a call at some point too, if you think any of these could use further discussion.

jackh726 committed:
Address some review comments

yaahc commented on src/expressions/block-expr.md:
```diff
@@ -63,6 +63,48 @@ assert_eq!(5, five);
 > As a control flow expression, if a block expression is the outer expression of an expression statement, the expected type is `()` unless it is followed immediately by a semicolon.
 
+r[expr.block.type.diverging]
+A block is itself considered to be [diverging](../divergence.md) if all reachable control flow paths contain a [diverging expression](../divergence.md#r-divergence.diverging-expressions), unless that expression is a place expression that is not read from.
```
place expression could also use a link: expressions.html#r-expr.place-value.place-memory-location

yaahc commented on src/divergence.md:
````diff
@@ -0,0 +1,37 @@
+>     4 => Default::default(),
+>     _ => return,
+> };
+> ```
````
is this addressed by the bit about type unification happening structurally below at https://github.com/rust-lang/reference/pull/2067/files#diff-11f5e85172f632195dd6a9a80daa852911fd3d66e788177a3bb39f6c5d7f21cfR27 ?

jackh726 replied:
It's a good observation. I've tried to be a little nuanced in my wording - for example, in the list of expression that I removed from your review, I had the wording "the following language constructs unconditonally produce a _diverging expression_ of the type `!`". There also still is the wording "Any expression of type [`!`](./types/never.md) is a _diverging expression_, but there are also diverging expressions which are not of type `!`"

yaahc commented on src/expressions/match-expr.md:
```diff
@@ -96,6 +96,27 @@ Every binding in each `|` separated pattern must appear in all of the patterns i
+The type of the overall `match` expression is the [least upper bound](../type-coercions.md#r-coerce.least-upper-bound) of the individual match arms.
+
+r[expr.match.empty]
+If there are no match arms, then the `match` expression is diverging and the type is [`!`](../types/never.md).
```
does this require a diverging scrutinee expression to compile?

probably worth adding an example either way

jackh726 replied:
I think it's useful to mention this, thoughts on this: https://github.com/rust-lang/reference/pull/2067/commits/fa9825d290f8a58ff7850a339f1955ff7fa14db2

jackh726 replied:
Changed this to be a `NOTE`.

jackh726 committed:
Use functions returning ! in tests

jackh726 committed:
Reviews from Jane

jackh726 committed:
Fix test - updating items.fn.implicit-return to be more specific

jackh726 committed:
Fix CI

nikomatsakis reviewed (commented):
Some drive-by comments.

jackh726 committed:
Add section about uninhabited types in divergence.md

jackh726 committed:
Address Niko's comments.

ehuss commented on src/items/functions.md:
```diff
@@ -56,7 +56,7 @@ r[items.fn.signature]
 
 r[items.fn.implicit-return]
-If the output type is not explicitly stated, it is the [unit type].
+If the output type is not explicitly stated, it is the [unit type]. However, if the block expression is not [diverging](../divergence.md), then the output type is instead [`!`](../types/never.md).
```
I don't understand this change. Can you explain what it is getting at?

We were initially thinking the `not` is wrong, but that also seems wrong since the implicit return type is still `()` for a function with a diverging block.

```rust
#![feature(never_type)]
pub fn f() {
    panic!("");
}

struct S {
    x: !
}

pub fn z() {
    S {x: f()}; // ERROR: expected `!`, found `()`
}
```

traviscross committed:
Make `diverging_place_not_read` subject to failure

This test is meant to show that matching a diverging place against the
wildcard pattern does not cause a read and therefore does not cause
the block to diverge.

This was shown by ascribing the return type of the function to unit.
But due to never-to-any coercion, that will always type check, even if
the read does take place.

Let's instead ascribe the function return type to never and assert
that the test must fail.  This way, the test would fail if the read
did occur.

ehuss committed:
Consistently document and link diverging expressions

This tries to add consistency to if, match, break, continue, return, and
loop in how the rule for divergence (and their type).

ehuss committed:
Fix misspelling in an example

Also consistently uses periods in comments.

ehuss committed:
Add note about use of an unstable nightly feature

This helps make it clear that this is using special syntax that is
normally not allowed.

Also, unhide `make` so it isn't so mysterious what it is.

ehuss committed:
Clarify subtlety about `if` example semicolon

ehuss committed:
Fix stray `>`

ehuss committed:
Remove use of nightly feature in an example

This particular example can be written with stable Rust, and I don't
think it detracts too much from what it is trying to illustrate.

ehuss committed:
Revert change to items.fn.implicit-return

The addition here isn't quite right. See discussion at
https://github.com/rust-lang/reference/pull/2067/changes#r2630177707.

As a later followup, we should document that the type of the function
block must be coercible into the return type of the function. (Or if
there is a more generalized way to state that.)

ehuss committed:
Clean up the links for expr.block.type.diverging

ehuss committed:
Clarify the wording for expr.block.type

Also clean up the links.

ehuss committed:
Clean up divergence chapter links

ehuss committed:
Rename rule expr.block.diverging

Rule names shouldn't be nested under other rules.

ehuss committed:
Rename expr.match.conditional

I'm not sure what happened here. The rule was originally
expr.match.type.diverging.conditional, then it was requested to change
to expr.match.diverging. I think that suggestion fits well here, so
let's go with it.

ehuss committed:
Consistently use periods in comments

ehuss committed:
Rework the divergence intro

This starts off the chapter with a direct definition of what a diverging
expression is (see discussion in
https://github.com/rust-lang/reference/pull/2067/changes#diff-11f5e85172f632195dd6a9a80daa852911fd3d66e788177a3bb39f6c5d7f21cfR10
about this definition).

This also adds a very simple example showing divergence. Maybe not the
greatest illustration, but I'd prefer to have at least a little
something.

ehuss committed:
Add back the list of divergence rules

This was asked to be changed or removed in
https://github.com/rust-lang/reference/pull/2067#discussion_r2475872395
and was removed in 5eef4bcc01f19c3ad4d0298df0337d0655538361, but I think
it is useful to include the links to the appropriate rules that are
relevant here.

This doesn't exactly duplicate any of the rules, it is just forwarding
links, and this is something we do all over the Reference.

I'm not 100% certain I want to keep this since if we list all expression
kinds, this could get long.

Also add a mention about panics. I'm not sure if putting this in a note
is really the right approach. However, we don't directly document
`panic!` yet. We are currently in discussion about adding documentation
for built-in macros.

ehuss committed:
Add divergence.never

Add a rule identifier to this particular statement. Also some slight
rewording for clarity.

ehuss committed:
Move propagation statement to a note

For the most part, this seemed to be restating the expr.block.diverging
rule. (At least, I couldn't determine if it was expressing something
distinct.)

This is an interesting property, but one that I think is already
covered. So, relegate it to a note linking to the relevant text.

jackh726 commented:
Other than the one comment above, the changes look fine and correct to me. I don't think there is really anything related to divergence that's left out of this PR.

jackh726 commented:
The following summarizes my understanding:
> 
> Already handled in this PR:
> ...
> The following are generally straightforward where they diverge if any of their operands diverge:
> But there are some that are more subtle:

This is a good shout-out, I didn't quite think about these. I think I focused solely on expressions that both diverge *and* have a type of `!`. Of course, `if` and `match` are also special so it was obvious to include those. Many of the remaining expressions (I think) fall out from the `diverging.intro` statement. Of course, explicitly stating these may still be useful. (I can also go through and respond to each one-by-one another time.)

> I also had a question about const-eval. My general understanding is that const eval is not allowed to diverge. Is that correct? I'm not sure how to express that or if my understanding is correct.

I'm not entirely sure, but it's not immediately clear to me that this is the case. Importantly, there is *divergence as it pertains to type checking* and const eval itself.

> The following has me a little confused. `const` blocks can pass `cargo check`, but fail with `cargo build`.
> 
> ```rust
> fn f() -> ! {
>     const { panic!() } // OK, seems const block has ! type
> }
> fn g() -> ! {
>     const { panic!() }; 
>     // ERROR, expected `!`, found `()`
> }
> ```
> 
> `g` confuses me because if the const block is diverging, then it should pass `cargo check` (like a non-const block). Perhaps divergence for those doesn't really matter if divergence in const eval is never allowed?


This has a pretty simple explanation: the former results in an expected type of `!` for the const block so the type of the const block is `!` - the latter has no expectation because of the semicolon, so the type of the block `!` fallbacks to `()`.

ehuss committed:
Add rule expr.loop.break-value.implicit-value

This lifts up this sentence (which was missing a rule), and clarifies
that this only applies to loops and labeled block expressions.

ehuss committed:
Specify the type of a labeled block expression

ehuss committed:
More clearly specify the type and divergence of a loop with break

ehuss committed:
Add some missing expressions for the LUB coercion list

ehuss committed:
Move link to link definition

ehuss committed:
Remove note about match divergence and `!`

I feel like this is repeating the rule `divergence.never`, and this
property is true of most expressions and I would prefer to not repeat it
(or inconsistently repeat it). For example, `Some(loop{})` diverges, but
would we also include a note in call expressions that it can diverge
even if its type is not `!`?

ehuss reviewed (approved):
Thanks, I really appreciate it! I'm going to go ahead and merge. I've opened a few issues to track followups.
